pkg/executor/wasm/executor.go

```diff
@@ -70,19 +81,23 @@ func (e *Executor) Start(ctx context.Context, request *executor.RunCommandReques
 		}
 	}
 
-	engineParams, err := wasmmodels.DecodeArguments(request.EngineParams)
-	if err != nil {
-		return fmt.Errorf("decoding wasm arguments: %w", err)
-	}
-
 	// Apply memory limits to the runtime. We have to do this in multiples of
 	// the WASM page size of 64kb, so round up to the nearest page size if the
 	// limit is not specified as a multiple of that.
 	engineConfig := wazero.NewRuntimeConfig().WithCloseOnContextDone(true)
 	if request.Resources.Memory > 0 {
-		const pageSize = 65536
-		pageLimit := request.Resources.Memory/pageSize + math.Min(request.Resources.Memory%pageSize, 1)
-		engineConfig = engineConfig.WithMemoryLimitPages(uint32(pageLimit))
+		requestedPages := request.Resources.Memory/WASM_PAGE_SIZE + math.Min(request.Resources.Memory%WASM_PAGE_SIZE, 1)
+		if requestedPages > WASM_MAX_PAGES_LIMIT {
+			err = fmt.Errorf("requested memory exceeds the wasm limit - %d > 4GB", request.Resources.Memory)
+			log.Err(err).Msgf("requested memory exceeds maximum limit: %d > %d", requestedPages, WASM_MAX_PAGES_LIMIT)
+			return err
+		}
+		engineConfig = engineConfig.WithMemoryLimitPages(uint32(requestedPages))
+	}
+
+	engineParams, err := wasmmodels.DecodeArguments(request.EngineParams)
```

Comment:
i had to reorder cuz my test was failing due to decoding wasm. Additionally it made more sense to check mem config first.